// @nx/react/src/generators/application/lib/bundlers/add-rspack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initRspack = initRspack;
exports.setupRspackConfiguration = setupRspackConfiguration;
exports.handleStyledJsxForRspack = handleStyledJsxForRspack;
const devkit_1 = require("@nx/devkit");
const pc = require("picocolors");
const versions_1 = require("../../../../utils/versions");
const maybe_js_1 = require("../../../../utils/maybe-js");
async function initRspack(tree, options, tasks) {
    const { rspackInitGenerator } = (0, devkit_1.ensurePackage)('@nx/rspack', versions_1.nxVersion);
    const rspackInitTask = await rspackInitGenerator(tree, {
        ...options,
        addPlugin: false,
        skipFormat: true,
    });
    tasks.push(rspackInitTask);
}
async function setupRspackConfiguration(tree, options, tasks) {
    const { configurationGenerator } = (0, devkit_1.ensurePackage)('@nx/rspack', versions_1.nxVersion);
    const rspackTask = await configurationGenerator(tree, {
        project: options.projectName,
        main: (0, devkit_1.joinPathFragments)(options.appProjectRoot, (0, maybe_js_1.maybeJs)({
            js: options.js,
            useJsx: true,
        }, `src/main.tsx`)),
        tsConfig: (0, devkit_1.joinPathFragments)(options.appProjectRoot, 'tsconfig.app.json'),
        target: 'web',
        newProject: true,
        framework: 'react',
    });
    tasks.push(rspackTask);
}
function handleStyledJsxForRspack(tasks, tree, options) {
    devkit_1.logger.warn(`${pc.bold('styled-jsx')} is not supported by ${pc.bold('Rspack')}. We've added ${pc.bold('babel-loader')} to your project, but using babel will slow down your build.`);
    tasks.push((0, devkit_1.addDependenciesToPackageJson)(tree, {}, { 'babel-loader': versions_1.babelLoaderVersion }));
    tree.write((0, devkit_1.joinPathFragments)(options.appProjectRoot, 'rspack.config.js'), (0, devkit_1.stripIndents) `
        const { composePlugins, withNx, withReact } = require('@nx/rspack');
        module.exports = composePlugins(withNx(), withReact(), (config) => {
          config.module.rules.push({
            test: /\\.[jt]sx$/i,
            use: [
              {
                loader: 'babel-loader',
                options: {
                  presets: ['@babel/preset-typescript'],
                  plugins: ['styled-jsx/babel'],
                },
              },
            ],
          });
          return config;
        });
        `);
}
